util/filesort/filesort.go

```diff
@@ -0,0 +1,515 @@
+		handle: dcod[fs.keySize+fs.valSize:][0].GetInt64(),
+	}, nil
+}
+
+func (fs *FileSorter) openAllFiles() error {
+	for _, fname := range fs.files {
+		fd, err := os.Open(fname)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		fs.fds = append(fs.fds, fd)
+	}
+	return nil
+}
+
+func (fs *FileSorter) closeAllFiles() error {
+	for _, fd := range fs.fds {
+		err := fd.Close()
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	err := os.RemoveAll(fs.tmpDir)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// Close terminates the input or output process and discards all remaining data.
+func (fs *FileSorter) Close() error {
+	if fs.closed {
+		return errors.New("FileSorter has been closed")
+	}
+	fs.buf = fs.buf[:0]
+	err := fs.closeAllFiles()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	fs.closed = true
+	return nil
+}
+
+// Output gets the next sorted row.
+func (fs *FileSorter) Output() ([]types.Datum, []types.Datum, int64, error) {
+	if fs.closed {
+		return nil, nil, 0, errors.New("FileSorter has been closed")
+	}
+	if len(fs.files) == 0 {
+		// No external files generated.
+		// Perform full in-memory sort directly.
+		r, err := fs.internalSort()
+		if err != nil {
+			return nil, nil, 0, errors.Trace(err)
+		} else if r != nil {
+			return r.key, r.val, r.handle, nil
+		} else {
+			return nil, nil, 0, nil
+		}
+	}
+	r, err := fs.externalSort()
+	if err != nil {
+		return nil, nil, 0, errors.Trace(err)
+	} else if r != nil {
+		return r.key, r.val, r.handle, nil
+	} else {
+		return nil, nil, 0, nil
+	}
+}
+
+// Perform full in-memory sort.
+func (fs *FileSorter) internalSort() (*comparableRow, error) {
+	if !fs.fetched {
+		sort.Sort(fs)
+		if fs.err != nil {
+			return nil, errors.Trace(fs.err)
+		}
+		fs.fetched = true
+	}
+	if fs.cursor < len(fs.buf) {
+		r := fs.buf[fs.cursor]
+		fs.cursor++
+		return r, nil
+	}
+	return nil, nil
+}
+
+// Perform external file sort.
+func (fs *FileSorter) externalSort() (*comparableRow, error) {
+	if !fs.fetched {
+		if len(fs.buf) > 0 {
+			err := fs.flushToFile()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+
+		heap.Init(fs.rowHeap)
+		if fs.rowHeap.err != nil {
+			return nil, errors.Trace(fs.rowHeap.err)
+		}
+
+		err := fs.openAllFiles()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		for id := range fs.fds {
+			row, err := fs.fetchNextRow(id)
+			if row == nil {
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				return nil, errors.New("file is empty")
+			}
+
+			im := &item{
+				index: id,
+				value: row,
+			}
+
+			heap.Push(fs.rowHeap, im)
+			if fs.rowHeap.err != nil {
+				return nil, errors.Trace(fs.rowHeap.err)
+			}
+		}
+
+		fs.fetched = true
+	}
+
+	if fs.rowHeap.Len() > 0 {
+		im := heap.Pop(fs.rowHeap).(*item)
+		if fs.rowHeap.err != nil {
+			return nil, errors.Trace(fs.rowHeap.err)
+		}
+
+		row, err := fs.fetchNextRow(im.index)
+		if row == nil {
+			if err != nil {
```

Comment:
My opinion here is that why you don't add code like line464. If you don't return here when row == nil && err == nil is true, I think line 488 can be removed.